src/librustc_mir/interpret/validity.rs

```diff
@@ -303,7 +303,7 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
         let (lo, hi) = layout.valid_range.clone().into_inner();
         let max_hi = u128::max_value() >> (128 - size.bits()); // as big as the size fits
         assert!(hi <= max_hi);
-        if lo == 0 && hi == max_hi {
+        if (lo == 0 && hi == max_hi) || (hi + 1 == lo) {
```

Comment:
We can unify these: `(hi + 1) % max_hi == lo`. Correct? That seems nicer.

Reply:
:laughing: I had that, we can't. It would need to be `x % (max_hi + 1)`, which doesn't work for `u128`. At which point we're getting way more complex than my simple two conditionals

Reply:
Okay. :)  Then please add a comment explaining exactly this.